lighthouse-core/lib/network-recorder.js

```diff
@@ -259,29 +294,87 @@ class NetworkRecorder extends EventEmitter {
     }
 
     switch (event.method) {
-      case 'Network.requestWillBeSent': return this.onRequestWillBeSent(event.params);
-      case 'Network.requestServedFromCache': return this.onRequestServedFromCache(event.params);
-      case 'Network.responseReceived': return this.onResponseReceived(event.params);
-      case 'Network.dataReceived': return this.onDataReceived(event.params);
-      case 'Network.loadingFinished': return this.onLoadingFinished(event.params);
-      case 'Network.loadingFailed': return this.onLoadingFailed(event.params);
-      case 'Network.resourceChangedPriority': return this.onResourceChangedPriority(event.params);
-      default: return;
+      case 'Network.requestWillBeSent':
+        return this.onRequestWillBeSent(event.params);
+      case 'Network.requestServedFromCache':
+        return this.onRequestServedFromCache(event.params);
+      case 'Network.responseReceived':
+        return this.onResponseReceived(event.params);
+      case 'Network.dataReceived':
+        return this.onDataReceived(event.params);
+      case 'Network.loadingFinished':
+        return this.onLoadingFinished(event.params);
+      case 'Network.loadingFailed':
+        return this.onLoadingFailed(event.params);
+      case 'Network.resourceChangedPriority':
+        return this.onResourceChangedPriority(event.params);
+      default:
+        return;
     }
   }
 
+  /**
+   * @param {string} requestId
+   * @return {NetworkRequest|undefined}
+   */
+  _findRealRequest(requestId) {
+    let request = this._recordsById.get(requestId);
+    if (!request) return undefined;
+
+    while (request.redirectDestination) {
+      request = request.redirectDestination;
+    }
+
+    return request;
+  }
+
   /**
    * Construct network records from a log of devtools protocol messages.
    * @param {LH.DevtoolsLog} devtoolsLog
    * @return {Array<LH.WebInspector.NetworkRequest>}
    */
   static recordsFromLogs(devtoolsLog) {
-    /** @type {Array<LH.WebInspector.NetworkRequest>} */
-    const records = [];
-    const nr = new NetworkRecorder(records);
-    devtoolsLog.forEach(message => {
-      nr.dispatch(message);
-    });
+    const networkRecorder = new NetworkRecorder();
+    // playback all the devtools messages to recreate network records
+    devtoolsLog.forEach(message => networkRecorder.dispatch(message));
+
+    // get out the list of records
+
+    const records = networkRecorder.getRecords();
+
+    // create a map of all the records by URL to link up initiator
+    const recordsByURL = new Map();
+    for (const record of records) {
+      if (recordsByURL.has(record.url)) continue;
+      recordsByURL.set(record.url, record);
+    }
+
+    // set the initiator and redirects array
```

Comment:
at first glance, it seems odd that this happens only when we're hydrating the records from logs. what about when records are created via `this._networkStatusMonitor.dispatch(event);` in driver?

Reply:
yeah it's a bit weird, AFAICT we actually only ever use the ones created by the recordsFromLogs, the ones created incrementally are just for the status notifications

https://github.com/GoogleChrome/lighthouse/blob/996c5d7ea19e1f7622fd4b2b2b85d6e60460a8d1/lighthouse-core/gather/gather-runner.js#L301
(and obvs the computed artifact uses this)

long-term though I hope to eliminate the initiatorRequest logic here entirely and move it to the page-dependency-graph since it's just CRC-based ones that need it right now

redirects can also move to direct usage of source/destination which I think becomes more explicit and clear

Reply:
is there a way to do this incrementally as (presumably) NetworkManager or whatever does? If the same thing is available in the page dependency graph I guess that's fine too, but at that point do we need NetworkRecords...

Reply:
I really, really want to nix these entirely so network request can be trivially serializable. Because of that, I don't really want to do unnecessary work to make them nice only to remove them

Reply:
> yeah it's a bit weird, AFAICT we actually only ever use the ones created by the recordsFromLogs, the ones created incrementally are just for the status notifications

ha, I think it was in anticipation of exactly this sort of situation. Since audits can only access network records created by `recordsFromLogs(devtoolsLogs)`, we wanted to make sure gatherers also saw exactly the same thing (in `afterPass()`) and so switched everything to using it.

So this impl and then migration from it sounds good. Nothing should be able to observe the difference since driver never actually inspects the records while listening to the quiet events